xds/src/main/java/com/linecorp/armeria/xds/client/endpoint/RetryStateFactory.java

```diff
@@ -0,0 +1,466 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.xds.client.endpoint;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.HttpClient;
+import com.linecorp.armeria.client.RefusedStreamException;
+import com.linecorp.armeria.client.UnprocessedRequestException;
+import com.linecorp.armeria.client.retry.Backoff;
+import com.linecorp.armeria.client.retry.RetryConfig;
+import com.linecorp.armeria.client.retry.RetryConfigBuilder;
+import com.linecorp.armeria.client.retry.RetryConfigMapping;
+import com.linecorp.armeria.client.retry.RetryDecision;
+import com.linecorp.armeria.client.retry.RetryRule;
+import com.linecorp.armeria.client.retry.RetryingClient;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.logging.RequestLog;
+import com.linecorp.armeria.common.logging.RequestLogProperty;
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
+import com.linecorp.armeria.xds.client.endpoint.RouteEntryMatcher.HeaderMatcherImpl;
+
+import io.envoyproxy.envoy.config.route.v3.HeaderMatcher;
+import io.envoyproxy.envoy.config.route.v3.RetryPolicy;
+import io.envoyproxy.envoy.config.route.v3.RetryPolicy.RateLimitedRetryBackOff;
+import io.envoyproxy.envoy.config.route.v3.RetryPolicy.ResetHeader;
+import io.envoyproxy.envoy.config.route.v3.RetryPolicy.RetryBackOff;
+import io.grpc.Status;
+
+final class RetryStateFactory {
+
+    private static final String REQUEST_HEADER_GRPC_RETRY_ON = "x-envoy-retry-grpc-on";
+    private static final String REQUEST_HEADER_RETRY_ON = "x-envoy-retry-on";
+    private static final String REQUEST_HEADER_MAX_RETRIES = "x-envoy-max-retries";
+    private static final String REQUEST_HEADER_RETRIABLE_STATUS_CODES = "x-envoy-retriable-status-codes";
+    private static final String REQUEST_HEADER_RETRIABLE_HEADER_NAMES = "x-envoy-retriable-header-names";
+    private static final String RESPONSE_HEADER_X_ENVOY_RATELIMITED = "x-envoy-ratelimited";
+    private static final String RESPONSE_HEADER_GRPC_STATUS = "grpc-status";
```

Comment:
nit: Should we create them using `HttpHeaderNames.of()`, which makes the header name comparison faster?